pkg/controllers/providers/instancetype/vmcapacitycache/controller.go

```diff
@@ -0,0 +1,105 @@
+	"go.uber.org/multierr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+)
+
+type Controller struct {
+	kubeClient           client.Client
+	instancetypeProvider *instancetype.DefaultProvider
+	initialized          bool
+}
+
+func NewController(kubeClient client.Client, instancetypeProvider *instancetype.DefaultProvider) *Controller {
+	return &Controller{
+		kubeClient:           kubeClient,
+		instancetypeProvider: instancetypeProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, r reconcile.Request) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "instancetypes.vmcapacitycache")
+	if !c.initialized {
+		return c.reconcileAllNodes(ctx)
+	}
+	if err := c.instancetypeProvider.UpdateVMCapacityCache(ctx, c.kubeClient, r.Name); err != nil {
+		return reconcile.Result{}, fmt.Errorf("updating vm capacity cache, %w", err)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) reconcileAllNodes(ctx context.Context) (reconcile.Result, error) {
+	nodeList := &corev1.NodeList{}
+	if err := c.kubeClient.List(ctx, nodeList, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{karpv1.NodeRegisteredLabelKey: "true"}),
+	}); err != nil {
+		return reconcile.Result{}, err
+	}
+	errs := make([]error, len(nodeList.Items))
+	for i, node := range nodeList.Items {
+		if err := c.instancetypeProvider.UpdateVMCapacityCache(ctx, c.kubeClient, node.Name); err != nil {
+			errs[i] = err
+		}
+	}
+	if err := multierr.Combine(errs...); err != nil {
+		return reconcile.Result{}, fmt.Errorf("updating vm capacity cache, %w", err)
+	}
+	c.initialized = true
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named("instancetypes.vmcapacitycache").
+		WatchesRawSource(singleton.Source()).
+		For(&corev1.Node{}).
+		WithEventFilter(predicate.Funcs{
+			// Only trigger reconciliation once a node becomes registered
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldNode, okOld := e.ObjectOld.(*corev1.Node)
+				newNode, okNew := e.ObjectNew.(*corev1.Node)
+				if !okOld || !okNew {
+					return false
+				}
+				return oldNode.Labels[karpv1.NodeRegisteredLabelKey] == "" && newNode.Labels[karpv1.NodeRegisteredLabelKey] == "true"
+			},
+			CreateFunc:  func(e event.CreateEvent) bool { return false },
+			DeleteFunc:  func(e event.DeleteEvent) bool { return false },
+			GenericFunc: func(e event.GenericEvent) bool { return false },
+		}).
```

Comment:
IIUC the purpose of the singleton source is to ensure we hydrate the cache on startup. Rather than having two separate reconciliation paths (one for individual Nodes, and one for startup where we list all nodes) what if we just reconcile against create events for registered Nodes. 
```suggestion
		For(&corev1.Node{}, builder.WithPredicates(predicate.TypedFuncs[client.Object]{
			// Only trigger reconciliation once a node becomes registered. This is an optimization to omit no-op reconciliations and reduce lock contention on the cache.
			UpdateFunc: func(e event.TypedUpdateEvent[client.Object]) bool {
			    if e.ObjectOld.GetLabels()[karpv1.NodeRegisteredLabelKey] != "" {
			        return false
			    }
			    return e.ObjectNew.GetLabels()[karpv1.NodeRegisteredLabelKey] == "true"
			},
			// Reconcile against all Nodes added to the cache in a registered state. This allows us to hydrate the cache on controller startup.
			CreateFunc:  func(e event.TypedCreateEvent[client.Object]) bool { 
			    return e.Object.GetLabels()[karpv1.NodeRegisteredLabelKey] == "true" 
			},
			DeleteFunc:  func(e event.TypedDeleteEvent[client.Object]) bool { return false },
			GenericFunc: func(e event.TypedGenericEvent[client.Object]) bool { return false },
		}).
```

Reply:
> Reconcile against all Nodes added to the cache in a registered state. This allows us to hydrate the cache on controller startup

Thanks for the education, I though it would still only react on _future_ node creation events.